### PR TITLE
Add preliminary support for reflection of Spectrum tables/columns.

### PR DESCRIPTION
**Pull request by briandailey**
Redshift Spectrum allows you to create tables that scan objects on S3.
This allows you to infer data types from the table.

Partially addresses #122.

## Todos
- [x] MIT compatible
- [x] Tests
- [x] Documentation
- [x] Updated CHANGES.rst

### CODE REVIEW

File: tests/test_reflection_views.py [hunk 1]
```diff
@@ -40,3 +40,25 @@ def test_late_binding_view_reflection(redshift_engine):
     view = Table('my_late_view', MetaData(),
                  autoload=True, autoload_with=redshift_engine)
     assert(len(view.columns) == 2)
+
+
+def test_late_binding_view_reflection(redshift_engine):
```
Comment (sburns) [inline tests/test_reflection_views.py]: @briandailey can you change this name to `test_spectrum_reflection` or something similar? name clashing in tests.

**Comment by jklukas** [comment]
LGTM; I just triggered tests to run against Redshift (which are flaky....); assuming tests looks good once we get a good run, this should be ready to merge.

**Comment by jklukas** [comment]
Tests are showing:

> E       ProgrammingError: (psycopg2.ProgrammingError) schema "spectrum" does not exist

**Comment by briandailey** [comment]
Hmm, I'm not sure how I would create a schema in these tests. Documentation on this is [here](https://docs.aws.amazon.com/redshift/latest/dg/c-getting-started-using-spectrum-create-external-table.html), but basically you need an IAM role that can access a bucket.

Snippet from docs:

```
create external schema spectrum 
from data catalog 
database 'spectrumdb' 
iam_role 'arn:aws:iam::123456789012:role/mySpectrumRole'
create external database if not exists;
```

I assume this would fail if it were an invalid IAM role. Any recommendations on the best way to write a test for this?

**Comment by jklukas** [comment]
I haven't touched the `bigcrunch` codebase we use for spinning up a cluster for these tests in a long time, but I think we'd need to make a change there to enable this. I had forgotten about the complexity of needing creds to hit S3 for Spectrum.

cc @graingert

**Comment by jklukas** [comment]
I just created a role and issued https://github.com/sqlalchemy-redshift/bigcrunch/pull/6 to make that role available to the transient test cluster.

**Comment by briandailey** [comment]
Ok, this looks dependent on this, so I'll wait until that merges. I'll try to keep an eye on it and make changes as necessary, but if someone beats me to it I'm not gonna be mad.

**Comment by MichaelVasseurBE** [comment]
Hi Team,
We are stuck on a critical project due to the lack of Redshift Spectrum support from the related SQLAlchemy driver. Do you have any timeline regarding the release of this implementation ?
Regards

**Comment by sburns** [comment]
I'm not sure what the difference is between these two Travis runs but the recent commit is passing, hopefully it can be merged soon. @MichaelVasseurBE, I know that @jklukas can't devote a ton of time to this project.

If all else fails you could install this branch yourself, something like `pip install https://github.com/briandailey/sqlalchemy-redshift/tarball/spectrum-support` should work.

**Comment by jklukas** [comment]
I personally am not going to have more time to devote to getting the testing infrastructure set up to validate this PR within the next week or two, so I'd agree that your best route at this point would be to `pip install` from the git branch where these changes live.

**Comment by MichaelVasseurBE** [comment]
Thank you guys, I'll give it a try !

**Comment by asokratis** [comment]
Looking forward to this as well. 

For now, the only workaround is creating clone tables and redirecting the metadata to those clones instead in redshift and then dropping them after that if they are not used (so you do not reach the redshift table quota limit)
```sql
DROP TABLE IF EXISTS redshift_schema.redshift_table;
CREATE TABLE redshift_schema.redshift_table 
AS ( SELECT * FROM external_schema.external_table WHERE 1 = 0);
# run sql alchemy code to extract redshift metadata
DROP TABLE redshift_schema.redshift_table;
```

**Comment by jklukas** [comment]
I just attempted to deploy our Redshift harness with the appropriate IAM role, but it failed due to using an old asyncio port of botocore that doesn't have the appropriate params for IAM role association. Rolling back now and thinking about how we might get around the problem.

**Comment by jklukas** [comment]
It looks like we'd need a complete rewrite of https://github.com/sqlalchemy-redshift/bigcrunch in order to avoid asyncio so we can use normal botocore with updated Redshift IAM methods.

That's unlikely to happen on a timescale of months, so I'm inclined to move forward with merging this.

@briandailey - If you mark the test that accesses S3 as skipped for now, with a comment about not being able to provision IAM creds for the test cluster, then I'll get this merged and deployed.

**Comment by graingert** [comment]
Does the new https://github.com/aio-libs/aiobotocore have the new IAM methods?

**Comment by jklukas** [comment]
> Does the new https://github.com/aio-libs/aiobotocore have the new IAM methods?

I had searched for redshift in there and found nothing, but perhaps that doesn't mean it's not supported? It may just mean that there aren't any tests for it.

**Comment by graingert** [comment]
It's because it pulls in all the code from botocore and most stuff is
auto-generated from Amazon's API manifest files
https://github.com/aio-libs/aiobotocore/blob/master/setup.py#L11

**Comment by jklukas** [comment]
aiobotocore is working! I've pushed changes to bigcrunch from a local branch. I will codify those in a PR later today.

The test here is now failing with:

> ProgrammingError: (psycopg2.ProgrammingError) schema "spectrum" does not exist

So we're getting close.

File: tests/test_reflection_views.py [hunk 2]
```diff
@@ -40,3 +40,25 @@ def test_late_binding_view_reflection(redshift_engine):
     view = Table('my_late_view', MetaData(),
                  autoload=True, autoload_with=redshift_engine)
     assert(len(view.columns) == 2)
+
+
+def test_spectrum_reflection(redshift_engine):
+    table_ddl = """create external table spectrum.sales(
```
Comment (jklukas) [inline tests/test_reflection_views.py]: Need to add here:

```
create external schema spectrum 
from data catalog 
database 'spectrumdb' 
iam_role 'arn:aws:iam::123456789012:role/mySpectrumRole'
create external database if not exists;
```

**Comment by jklukas** [comment]
I made some additional permissions changes to allow more progress here. I created a branch, see https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/compare/master...spectrum-support?expand=1

It's now failing with `KeyError: 'spectrum.sales'`, so it looks like there's something broken in the implementation. I've run out of time to look at this, so hopefully somebody can pick up from that branch and get this across the finish line.

**Comment by graingert** [comment]
@briandailey any updates on this?

**Comment by Brooke-white** [comment]
closing as #263 was merged

**Pull request closed without merging**